### PR TITLE
[MISC] Move seqan3::views::zip to seqan3/utility/views/

### PR DESCRIPTION
**Pull request by simonsasse**
Resolves: https://github.com/seqan/product_backlog/issues/357

### CODE REVIEW

File: test/performance/alignment/edit_distance_unbanded_benchmark.cpp [hunk 1]
```diff
@@ -15,7 +15,7 @@
 
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/alignment/edit_distance_unbanded_benchmark.cpp:18]: same here

File: test/performance/alignment/global_affine_alignment_benchmark.cpp [hunk 2]
```diff
@@ -16,7 +16,7 @@
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/alphabet/aminoacid/aa20.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/alignment/global_affine_alignment_benchmark.cpp:19]: same here

File: test/performance/alignment/global_affine_alignment_parallel_benchmark.cpp [hunk 3]
```diff
@@ -18,7 +18,7 @@
 #include <seqan3/alphabet/aminoacid/aa20.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/alignment/global_affine_alignment_parallel_benchmark.cpp:21]: same here

File: test/performance/alignment/global_affine_alignment_simd_benchmark_template.hpp [hunk 4]
```diff
@@ -16,7 +16,7 @@
 #include <seqan3/alphabet/aminoacid/aa20.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/alignment/global_affine_alignment_simd_benchmark_template.hpp:19]: same here

File: test/performance/alignment/local_affine_alignment_benchmark.cpp [hunk 5]
```diff
@@ -16,7 +16,7 @@
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/alphabet/aminoacid/aa20.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/alignment/local_affine_alignment_benchmark.cpp:19]: same here

File: test/performance/search/dream_index/interleaved_bloom_filter_benchmark.cpp [hunk 6]
```diff
@@ -8,7 +8,7 @@
 #include <benchmark/benchmark.h>
 
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/search/dream_index/interleaved_bloom_filter_benchmark.cpp:11]: same here

File: test/performance/utility/simd/views/to_simd_benchmark.cpp [hunk 7]
```diff
@@ -18,7 +18,7 @@
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/range/container/aligned_allocator.hpp>
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/performance/utility/simd/views/to_simd_benchmark.cpp:21]: same here

File: test/unit/alignment/pairwise/alignment_configurator_test.cpp [hunk 8]
```diff
@@ -16,7 +16,7 @@
 #include <seqan3/alignment/pairwise/detail/type_traits.hpp>
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/range/views/chunk.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/alignment/pairwise/alignment_configurator_test.cpp:19]: same here

File: test/unit/alignment/pairwise/pairwise_alignment_collection_callback_test_template.hpp [hunk 9]
```diff
@@ -14,7 +14,7 @@
 
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/range/views/to_char.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/alignment/pairwise/pairwise_alignment_collection_callback_test_template.hpp:17]: same here

File: test/unit/alignment/pairwise/pairwise_alignment_collection_test_template.hpp [hunk 10]
```diff
@@ -15,7 +15,7 @@
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/range/views/to_char.hpp>
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/alignment/pairwise/pairwise_alignment_collection_test_template.hpp:18]: same here

File: test/unit/core/algorithm/detail/algorithm_executor_blocking_test.cpp [hunk 11]
```diff
@@ -13,7 +13,7 @@
 
 #include <seqan3/core/algorithm/detail/algorithm_executor_blocking.hpp>
 #include <seqan3/range/views/persist.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/core/algorithm/detail/algorithm_executor_blocking_test.cpp:16]: same here

File: test/unit/core/algorithm/detail/execution_handler_template.hpp [hunk 12]
```diff
@@ -17,7 +17,7 @@
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/range/views/chunk.hpp>
 #include <seqan3/range/views/type_reduce.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/core/algorithm/detail/execution_handler_template.hpp:20]: same here

File: test/unit/io/detail/detail_record_test.cpp [hunk 13]
```diff
@@ -12,7 +12,7 @@
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/alphabet/quality/phred42.hpp>
 #include <seqan3/io/detail/record.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
 
 #include <seqan3/test/expect_same_type.hpp>
```
Comment (Irallia) [inline test/unit/io/detail/detail_record_test.cpp:18]: I think it is nicer, if the seqan/test includes are seperated, as here. But I found no rule for this. @eseiler?
https://github.com/seqan/seqan3/wiki/File-structure-and-naming

File: test/unit/io/sequence_file/sequence_file_format_test_template.hpp [hunk 14]
```diff
@@ -20,7 +20,7 @@
 #include <seqan3/io/sequence_file/input.hpp>
 #include <seqan3/io/sequence_file/output.hpp>
 #include <seqan3/range/views/convert.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/io/sequence_file/sequence_file_format_test_template.hpp:23]: same here

File: test/unit/range/views/get_test.cpp [hunk 15]
```diff
@@ -18,7 +18,7 @@
 #include <seqan3/range/views/get.hpp>
 #include <seqan3/range/views/complement.hpp>
 #include <seqan3/range/views/to_char.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
 #include <seqan3/std/ranges>
```
Comment (Irallia) [inline test/unit/range/views/get_test.cpp:22]: And can you move the seqan/std/ include to the iostream as it is a std include?

File: test/unit/range/views/single_pass_input_test.cpp [hunk 16]
```diff
@@ -15,7 +15,7 @@
 
 #include <seqan3/range/views/single_pass_input.hpp>
 #include <seqan3/range/views/persist.hpp>
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
```
Comment (Irallia) [inline test/unit/range/views/single_pass_input_test.cpp:18]: same here

File: test/unit/range/views/zip_test.cpp [hunk 17]
```diff
@@ -13,7 +13,7 @@
 
 #include <gtest/gtest.h>
 
-#include <seqan3/range/views/zip.hpp>
+#include <seqan3/utility/views/zip.hpp>
 #include <seqan3/std/algorithm>
 #include <seqan3/std/ranges>
```
Comment (Irallia) [inline test/unit/range/views/zip_test.cpp:18]: and these std includes aswell, thank you!

**Review by Irallia — CHANGES_REQUESTED** [review]
Could you please recheck the lexicographical order of all the includes? Sry I opened a lot of requests here. And there is a missing endline.
Thanks!

**Comment by simonsasse** [comment]
@Irallia I'm so sorry I completely forgot the reordering ! Sorry for the unnecessary work !

File: include/seqan3/range/views/zip.hpp [hunk 18]
```diff
@@ -7,20 +7,12 @@
...
-#include <seqan3/std/ranges>
-
-namespace seqan3::views
-{
-
-//TODO reimplement me
-using ::ranges::views::zip;
-
-} // namespace seqan3::views
+SEQAN3_DEPRECATED_HEADER("This header is deprecated and will be removed in SeqAn-3.1.0; Please #include <seqan3/utility/views/zip.hpp> instead.")
```
[the remark this answers is not in the document]
Comment (simonsasse) [reply include/seqan3/range/views/zip.hpp]: I am not able to see any end lines on GitHub in any seqan file. Is it possible that github is not displaying them?
In the new commit, I checked and added end lines and they are still not visible.

**Comment by simonsasse** [comment]
I separate the `seqan3/test/*` includes and reordered the includes. locally I can see the new lines at the end of the files, but not here on GitHub.  Thanks for the review. :)

**Review by Irallia — APPROVED** [review]
Thanks for the changes, now it looks good for me. 👍

File: test/unit/io/detail/detail_record_test.cpp (hunk 13 above)
Comment (eseiler) [reply test/unit/io/detail/detail_record_test.cpp:18]: Every include that starts with `#include <seqan3/` should go into the same block, so `utility` comes after `test`

File: include/seqan3/utility/views/zip.hpp [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+#pragma once
+
+#include <range/v3/view/zip.hpp>
+
+#include <seqan3/core/platform.hpp>
+#include <seqan3/std/ranges>
+
+namespace seqan3::views
```
Comment (eseiler) [inline include/seqan3/utility/views/zip.hpp:20]:
```suggestion
#pragma once

#include <seqan3/std/ranges>

#include <range/v3/view/zip.hpp>

#include <seqan3/core/platform.hpp>

namespace seqan3::views
```

**Review by eseiler — CHANGES_REQUESTED** [review]
Please re-put the test includes into the seqan3 include block

File: include/seqan3/utility/views/zip.hpp (hunk 19 above)
Comment (Irallia) [reply include/seqan3/utility/views/zip.hpp:20]: > Please re-put the test includes into the seqan3 include block

Sry for the extra work. Thanks for checking enrico :)

File: test/unit/alignment/pairwise/pairwise_alignment_collection_callback_test_template.hpp [hunk 20]
```diff
@@ -14,8 +14,9 @@
 
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/range/views/to_char.hpp>
-#include <seqan3/range/views/zip.hpp>
 #include <seqan3/test/expect_range_eq.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
 
 #include "fixture/alignment_fixture.hpp"
```
Comment (eseiler) [inline test/unit/alignment/pairwise/pairwise_alignment_collection_callback_test_template.hpp:21]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

#include "fixture/alignment_fixture.hpp"
```

File: test/unit/alignment/pairwise/pairwise_alignment_collection_test_template.hpp [hunk 21]
```diff
@@ -15,8 +15,9 @@
 #include <seqan3/alignment/pairwise/align_pairwise.hpp>
 #include <seqan3/range/views/to_char.hpp>
 #include <seqan3/range/views/to.hpp>
-#include <seqan3/range/views/zip.hpp>
 #include <seqan3/test/expect_range_eq.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
 
 #include "fixture/alignment_fixture.hpp"
```
Comment (eseiler) [inline test/unit/alignment/pairwise/pairwise_alignment_collection_test_template.hpp:22]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

#include "fixture/alignment_fixture.hpp"
```

File: test/unit/core/algorithm/detail/algorithm_executor_blocking_test.cpp [hunk 22]
```diff
@@ -13,8 +13,9 @@
 
 #include <seqan3/core/algorithm/detail/algorithm_executor_blocking.hpp>
 #include <seqan3/range/views/persist.hpp>
-#include <seqan3/range/views/zip.hpp>
 #include <seqan3/test/pretty_printing.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
```
Comment (eseiler) [inline test/unit/core/algorithm/detail/algorithm_executor_blocking_test.cpp:19]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

```

File: test/unit/io/detail/detail_record_test.cpp [hunk 23]
```diff
@@ -12,9 +12,9 @@
 #include <seqan3/alphabet/nucleotide/dna4.hpp>
 #include <seqan3/alphabet/quality/phred42.hpp>
 #include <seqan3/io/detail/record.hpp>
-#include <seqan3/range/views/zip.hpp>
-
 #include <seqan3/test/expect_same_type.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
```
Comment (eseiler) [inline test/unit/io/detail/detail_record_test.cpp:18]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

```

File: test/unit/range/views/get_test.cpp [hunk 24]
```diff
@@ -18,9 +19,9 @@
 #include <seqan3/range/views/get.hpp>
 #include <seqan3/range/views/complement.hpp>
 #include <seqan3/range/views/to_char.hpp>
-#include <seqan3/range/views/zip.hpp>
-#include <seqan3/std/ranges>
 #include <seqan3/test/expect_range_eq.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
```
Comment (eseiler) [inline test/unit/range/views/get_test.cpp:25]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

```

File: test/unit/range/views/single_pass_input_test.cpp [hunk 25]
```diff
@@ -15,8 +15,9 @@
 
 #include <seqan3/range/views/single_pass_input.hpp>
 #include <seqan3/range/views/persist.hpp>
-#include <seqan3/range/views/zip.hpp>
 #include <seqan3/test/expect_same_type.hpp>
+#include <seqan3/utility/views/zip.hpp>
+
```
Comment (eseiler) [inline test/unit/range/views/single_pass_input_test.cpp:21]:
```suggestion
#include <seqan3/utility/views/zip.hpp>

```

**Review by eseiler — APPROVED** [review]

**Pull request merged**